core/storage/base_model/gae_models.py

```diff
@@ -263,6 +264,21 @@ def get_field_names_for_takeout(cls) -> Dict[str, str]:
         """
         return {}
 
+    @overload
+    @classmethod
+    def get(
+        cls: Type[SELF_BASE_MODEL],
+        entity_id: str,
+    ) -> SELF_BASE_MODEL: ...
+
+    @overload
+    @classmethod
+    def get(
+        cls: Type[SELF_BASE_MODEL],
+        entity_id: str,
+        strict: bool = False
+    ) -> Optional[SELF_BASE_MODEL]: ...
```

Comment:
Does this work? It seems a bit weird I would expect to see `Literal[False]` and `Literal[True]`, etc.

Reply:
if we use Literals as follows:
```      
    @overload
    @classmethod
    def get(
        cls: Type[SELF_BASE_MODEL],
        entity_id: str,
        strict: Literal[True]
    ) -> Optional[SELF_BASE_MODEL]: ...
```
then mypy assumes that the type of strict argument is True ( not bool ).
So, if we call the get method like `get('id',  True)`, this is ok.
But if we call get method like this
```
    @classmethod
    def get( # type: ignore[override]
        cls, user_id: str, story_id: str, strict: bool = True
    ) -> Optional[StoryProgressModel]:
        .
        .
        .
        return super(StoryProgressModel, cls).get(
            instance_id, strict=strict)
```
Then according to mypy, it is wrong. because here we are passing bool ( not True ), there are many places in the codebase where we are following the second scenario. Some of them are:

https://github.com/oppia/oppia/blob/cf1826085ae5ca3a774aa64d4329b4a8ebbb483d/core/storage/statistics/gae_models.py#L124

https://github.com/oppia/oppia/blob/cf1826085ae5ca3a774aa64d4329b4a8ebbb483d/core/storage/user/gae_models.py#L2002

However, this thing is more deeply explained in the official doc's first example ( https://mypy.readthedocs.io/en/stable/literal_types.html#literal-types )

Why I'm not doing the overloads as follows?

```
    @overload
    @classmethod
    def get(
        cls: Type[SELF_BASE_MODEL],
        entity_id: str,
        strict: bool = True
    ) -> SELF_BASE_MODEL: ...

    @overload
    @classmethod
    def get(
        cls: Type[SELF_BASE_MODEL],
        entity_id: str,
        strict: bool = False
    ) -> Optional[SELF_BASE_MODEL]: ...

    @classmethod
    def get(
        cls: Type[SELF_BASE_MODEL],
        entity_id: str,
        strict: bool = True
    ) -> Optional[SELF_BASE_MODEL]:
        implementation........
```

Because the types of arguments of signature 1 match the types of arguments of signature 2 and due to this mypy is not able to differentiate between these overloads. Thus to overcome this issue, I removed `strict: bool = True` from the first overload because `strict: bool = True` is the default behavior.

Reply:
Ok, thanks for explanation. But thinking about it wouldn't it still be useful to add the overloads with `Literal` so that in cases where the variable is set explicitly we can return the correct type?

Also I think we should add overloads for the `get` functions in child classes of this class.

Reply:
overloads for the get functions are added. PTAL!